File: TController/src/DataProcessor/reconstructor.py
"""
AS7341 10 通道 → 全光谱重建。

使用 ams-OSRAM 官方 Golden Device 校准矩阵将 10 通道
(F1–F8, Clear, NIR) 原始 ADC 值重建为 380–1100 nm
连续全光谱（1 nm 步长，721 点）。

重建流程::

    corrected = factor × max(raw − offset, 0)
    spectrum[λ] = Σ factor[ch] × corrected[ch] × matrix[λ, ch]
"""

from __future__ import annotations

import os
from collections.abc import Sequence

import numpy as np

from DataProcessor._path import CALIBRE_PATH

_DATA_PATH = CALIBRE_PATH

_lazy: dict | None = None


def _load() -> dict:
    global _lazy
    if _lazy is not None:
        return _lazy
    if not os.path.isfile(_DATA_PATH):
        raise FileNotFoundError(f"光谱校准数据未找到: {_DATA_PATH}")
    data = np.load(_DATA_PATH, allow_pickle=True)
    _lazy = {
        k.replace("spectral_", ""): v
        for k, v in data.items()
        if k.startswith("spectral_")
    }
    return _lazy


def is_available() -> bool:
    """检查校准数据文件是否存在。"""
    return os.path.isfile(_DATA_PATH)


def get_wavelengths() -> np.ndarray:
    """返回波长数组 (380–1100 nm, 1 nm 步长)。"""
    return _load()["wavelengths"].copy()


def reconstruct(
    raw_values: Sequence[float] | np.ndarray,
    offsets: np.ndarray | None = None,
    factors: np.ndarray | None = None,
) -> tuple[np.ndarray, np.ndarray]:
    """从 10 通道原始 ADC 值重建全光谱。

    参数
    ----
    raw_values:
        10 元素序列 [F1, F2, F3, F4, F5, F6, F7, F8, Clear, NIR]。
    offsets:
        每通道暗电流/偏移。默认 → Golden Device 参考值。
    factors:
        每通道校正系数。默认 → Golden Device 参考值。

    返回
    ----
    (wavelengths, spectrum)
        两个一维 NumPy 数组 (len = 721)。
        wavelengths — 纳米波长的点
        spectrum   — 相对强度值 (a.u.)
    """
    cal = _load()
    raw = np.asarray(raw_values, dtype=np.float64)
    if raw.shape != (10,):
        raise ValueError(f"需要 10 通道数据，传入形状为 {raw.shape}")
    if np.any(raw < 0):
        raise ValueError("原始通道值不应包含负数")
    if np.any(np.isnan(raw)) or np.any(np.isinf(raw)):
        raise ValueError("原始通道值包含 NaN 或 Inf")

    ofs = cal["offsets"] if offsets is None else np.asarray(offsets, dtype=np.float64)
    fac = cal["factors"] if factors is None else np.asarray(factors, dtype=np.float64)
    
    if offsets is not None and (ofs.shape != (10,) or np.any(np.isnan(ofs))):
        raise ValueError("offsets 必须为长度 10 的有效数组")
    if factors is not None and (fac.shape != (10,) or np.any(fac <= 0) or np.any(np.isnan(fac))):
        raise ValueError("factors 必须为长度 10 的正数数组")

    corrected = fac * np.maximum(raw - ofs, 0.0)
    spectrum = np.maximum(
        cal["matrix"] @ corrected, 0.0
    )  # (721, 10) @ (10,) → (721,)

    return cal["wavelengths"].copy(), spectrum

File: TController/crates/controller-core/src/processing/reconstructor.rs
//! AS7341 10 通道 → 全光谱重建 — Python `reconstructor.py` 的移植。
//!
//! ams-OSRAM Golden Device 校准矩阵（`calibre.npz`）：
//! `corrected = factor × max(raw − offset, 0)`，
//! `spectrum[λ] = Σ_ch corrected[ch] × matrix[λ, ch]`（380–1100 nm，1 nm 步长，721 点）。

use std::path::{Path, PathBuf};

use ndarray::Array2;
use ndarray_npy::NpzReader;
use thiserror::Error;

/// 重建错误。
#[derive(Debug, Error)]
pub enum ReconError {
    #[error("光谱校准数据未找到: {0}")]
    FileNotFound(PathBuf),
    #[error("校准数据读取失败: {0}")]
    Read(String),
    #[error("需要 10 通道数据，传入长度为 {0}")]
    BadChannelCount(usize),
    #[error("原始通道值不应包含负数")]
    NegativeValue,
    #[error("原始通道值包含 NaN 或 Inf")]
    NonFinite,
}

/// Golden Device 校准数据（数值部分）。
#[derive(Debug, Clone)]
pub struct Reconstructor {
    /// 380–1100 nm（721 点）。
    pub wavelengths: Vec<f64>,
    /// (721, 10) 重建矩阵。
    pub matrix: Array2<f64>,
    pub offsets: Vec<f64>,
    pub factors: Vec<f64>,
}

impl Reconstructor {
    /// 从 `calibre.npz` 加载数值键（对象键如电极元数据不在此层处理）。
    pub fn load(path: &Path) -> Result<Self, ReconError> {
        if !path.is_file() {
            return Err(ReconError::FileNotFound(path.to_path_buf()));
        }
        let file = std::fs::File::open(path)
            .map_err(|e| ReconError::Read(format!("{}: {e}", path.display())))?;
        let mut npz = NpzReader::new(file).map_err(|e| ReconError::Read(e.to_string()))?;

        let read_f64_1d =
            |npz: &mut NpzReader<std::fs::File>, key: &str| -> Result<Vec<f64>, ReconError> {
                let arr: ndarray::Array1<f64> = npz
                    .by_name(key)
                    .map_err(|e| ReconError::Read(format!("{key}: {e}")))?;
                Ok(arr.to_vec())
            };

        let matrix: ndarray::Array2<f64> = npz
            .by_name("spectral_matrix")
            .map_err(|e| ReconError::Read(format!("spectral_matrix: {e}")))?;
        let wl_i32: ndarray::Array1<i32> = npz
            .by_name("spectral_wavelengths")
            .map_err(|e| ReconError::Read(format!("spectral_wavelengths: {e}")))?;
        let offsets = read_f64_1d(&mut npz, "spectral_offsets")?;
        let factors = read_f64_1d(&mut npz, "spectral_factors")?;

        Ok(Self {
            wavelengths: wl_i32.iter().map(|&v| v as f64).collect(),
            matrix,
            offsets,
            factors,
        })
    }

    /// 按开发/打包规则探测 `calibre.npz`：
    /// 1. 环境变量 `AUTOTITRATOR_CALIBRE`；
    /// 2. 可执行文件同级目录；
    /// 3. 开发模式：最终 workspace 的 `TController/data/`（编译期锚定）。
    pub fn discover() -> Result<(Self, PathBuf), ReconError> {
        let candidates: Vec<PathBuf> = {
            let mut v = Vec::new();
            if let Ok(p) = std::env::var("AUTOTITRATOR_CALIBRE") {
                v.push(PathBuf::from(p));
            }
            v.push(
                std::env::current_exe()
                    .ok()
                    .and_then(|e| e.parent().map(|d| d.join("calibre.npz")))
                    .unwrap_or_default(),
            );
            v.push(PathBuf::from(concat!(
                env!("CARGO_MANIFEST_DIR"),
                "/../../../TController/data/calibre.npz"
            )));
            v
        };
        for path in candidates {
            if path.is_file() {
                return Ok((Self::load(&path)?, path));
            }
        }
        Err(ReconError::FileNotFound(PathBuf::from("calibre.npz")))
    }

    /// 从 10 通道原始 ADC 值重建全光谱，返回 `(波长, 谱)`。
    pub fn reconstruct(&self, raw: &[f64]) -> Result<(Vec<f64>, Vec<f64>), ReconError> {
        if raw.len() != 10 {
            return Err(ReconError::BadChannelCount(raw.len()));
        }
        if raw.iter().any(|&v| v < 0.0) {
            return Err(ReconError::NegativeValue);
        }
        if !raw.iter().all(|v| v.is_finite()) {
            return Err(ReconError::NonFinite);
        }
        let corrected: Vec<f64> = raw
            .iter()
            .zip(&self.offsets)
            .zip(&self.factors)
            .map(|((&r, &o), &f)| f * (r - o).max(0.0))
            .collect();
        // spectrum = max(matrix @ corrected, 0)
        let mut spectrum = vec![0.0f64; self.matrix.nrows()];
        for (row_out, row) in spectrum.iter_mut().zip(self.matrix.rows()) {
            *row_out = row
                .iter()
                .zip(&corrected)
                .map(|(&m, &c)| m * c)
                .sum::<f64>()
                .max(0.0);
        }
        Ok((self.wavelengths.clone(), spectrum))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn discover_and_reconstruct_roundtrip() {
        let (rec, path) = Reconstructor::discover().expect("calibre.npz");
        assert!(path.is_file());
        assert_eq!(rec.wavelengths.len(), 721);
        assert_eq!(rec.matrix.dim(), (721, 10));

        let raw = [1000.0f64; 10];
        let (wls, spectrum) = rec.reconstruct(&raw).unwrap();
        assert_eq!(wls.len(), spectrum.len());
        assert!(spectrum.iter().all(|v| v.is_finite() && *v >= 0.0));

        assert!(matches!(
            rec.reconstruct(&[1.0; 10][..9]),
            Err(ReconError::BadChannelCount(9))
        ));
        assert!(matches!(
            rec.reconstruct(&[-1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]),
            Err(ReconError::NegativeValue)
        ));
    }
}
